Add support for processing accelerators hardware

mlorenzofr opened the pull request:
This PR adds support for detection of processing accelerators to the library. Please take a look.
This fixes #383

jaypipes reviewed (commented):
@mlorenzofr I absolutely love your ambition :) And overall, the quality of your PR is outstanding, thank you!

That said, I'm going to request that you remove the filtering functionality from this PR and add the filtering functionality in a separate follow-up PR? That way I can evaluate that separately. The reason is because none of the other `ghw` functions have a filter parameter so adding one just for the accelerator package would be mis-aligned.

mlorenzofr commented:
I was thinking... this PR will detect processing accelerators hardware, but there are hardware from some vendors that are not categorized as processing accelerators, but as 3D controllers. Should we add that hardware in the default detection until the custom filters feature is added?

jaypipes commented on pkg/accelerator/accelerator.go:
```diff
@@ -0,0 +1,84 @@
+	Address string `json:"address"`
+	// pointer to a PCIDevice struct that describes the vendor and product
+	// model, etc
+	Pci *pci.Device `json:"pci"`
```
Let's name this field `PCIDevice` (and json of `pci_device`) please.

jaypipes commented on pkg/accelerator/accelerator.go:
```diff
@@ -0,0 +1,84 @@
+	"github.com/jaypipes/ghw/pkg/pci"
+)
+
+type AcceleratorCard struct {
```
Instead of calling it `AcceleratorCard`, what do you think about `AcceleratorDevice`? Not all accelerators are necessarily "cards", after all, and "device" is a bit more generic...

mlorenzofr replied:
good idea, _device_ is much more generic and appropriate than _card_

jaypipes commented on pkg/accelerator/accelerator.go:
```diff
@@ -0,0 +1,84 @@
+	)
+}
+
+// simple private struct used to encapsulate gpu information in a top-level
+// "gpu" YAML/JSON map/object key
```
```suggestion
// simple private struct used to encapsulate accelerator information in a top-level
// "accelerator" YAML/JSON map/object key
```

jaypipes reviewed (commented):
@mlorenzofr thank you sir! a few suggested changes inline :)

mlorenzofr commented:
I've added all the suggestions, I loved them :heart: . Thanks a lot for the contributions. :pray: 

Additionally, since we've added discovery for more hardware, I've added an additional test for Nvidia cards.

jaypipes commented on pkg/accelerator/accelerator.go:
```diff
@@ -0,0 +1,84 @@
+
+type Info struct {
+	ctx                *context.Context
+	AcceleratorDevices []*AcceleratorDevice `json:"devices"`
```
One tiny request... let's call the field `Devices` instead of `AcceleratorDevices` since the "Accelerator" part is assumed.

mlorenzofr commented:
I forgot to update the documentation, I fixed it the commit 58df8e540fc8d4687749465a5f83410665bb8f1b

jaypipes reviewed (approved):
Fantastic! :) Thank you so much @mlorenzofr!

mlorenzofr commented on pkg/accelerator/accelerator_linux.go:
```diff
@@ -0,0 +1,72 @@
+		class := dev.Class.ID
+		subclass := dev.Subclass.ID
+		if subclasses, ok := acceleratorPCIClasses[class]; ok {
+			if slices.Contains(subclasses, subclass) {
```
So sorry! :pray:
I didn't realize, I just pushed the fix

jaypipes replied:
@mlorenzofr you have no reason to apologize at all!

jaypipes reviewed (approved):
Eccezionale :)

jaypipes commented:
@mlorenzofr welcome to the `ghw` Contributor roll! :) thank you so much for this contribution!

ffromani commented:
Nice work indeed!
(why I'm getting notifications so late?)

jaypipes commented:
> Nice work indeed! (why I'm getting notifications so late?)

LOL, probably because I [made a comment in Italian earlier](https://github.com/jaypipes/ghw/pull/385#pullrequestreview-2368734574) and that triggered your notifications :P

ffromani commented on README.md:
```diff
@@ -945,6 +946,60 @@ information
+)
+
+func main() {
+	accel, err := ghw.Accelerator()
```
do we want to extend (maybe later?) this API to accept additive user-provided filters?
I'm referring to the wide variety of devices out there and how they present themselves on PCI bus.
By "additive" filter I mean
- users cannot rule out a device `ghw` is really sure it's an accelerator (this means `ghw` must be conservative in the core logic, which I think it's good anyway)
- users can opt-in devices if they know better than `ghw` for whatever reasons (maybe old `ghw` version, maybe rare/custom devices)

ffromani replied:
perfect! I'll be on the lookout this time!
